odoo/addons/base/models/qweb.py

```diff
@@ -864,832 +658,553 @@ def _compile_static_node(self, el, options):
+
+        return code
+
+    def _compile_directive_set(self, el, options, indent):
         varname = el.attrib.pop('t-set')
+        directive = f't-set="{varname}"'
+        code = self._flushText(options, indent)
 
         if 't-value' in el.attrib:
-            value = self._compile_expr(el.attrib.pop('t-value') or 'None')
+            if varname == '0':
+                raise ValueError('t-set="0" should not contains t-value or t-valuef')
+            expr = el.attrib.pop('t-value') or 'None'
+            directive = f'{directive} t-value="{expr}"'
+            expr = self._compile_expr(expr)
         elif 't-valuef' in el.attrib:
-            value = self._compile_format(el.attrib.pop('t-valuef'))
+            if varname == '0':
+                raise ValueError('t-set="0" should not contains t-value or t-valuef')
+            exprf = el.attrib.pop('t-valuef')
+            directive = f'{directive} t-valuef="{exprf}"'
+            expr = self._compile_format(exprf)
         else:
             # set the content as value
-            body = self._compile_directive_content(el, options)
-            if body:
-                def_name = self._create_def(options, body, prefix='set', lineno=el.sourceline)
-                # ''.join($varset)
-                joined = ast.Call(
-                    func=ast.Attribute(value=ast.Str(u''), attr='join', ctx=ast.Load()),
-                    args=[ast.Name(id='content', ctx=ast.Load())], keywords=[]
-                )
-                # Markup(_): rendering of content is considered safe
-                safe = ast.Call(
-                    func=ast.Name(id='Markup', ctx=ast.Load()),
-                    args=[joined], keywords=[]
-                )
-                return [
-                    # content = []
-                    ast.Assign(
-                        targets=[ast.Name(id='content', ctx=ast.Store())],
-                        value=ast.List(elts=[], ctx=ast.Load())
-                    ),
-                    # set(self, $varset.append)
-                    ast.Expr(self._call_def(
-                        def_name,
-                        append=ast.Attribute(
-                            value=ast.Name(id='content', ctx=ast.Load()),
-                            attr='append',
-                            ctx=ast.Load()
-                        )
-                    )),
-                    # $varset = _
-                    ast.Assign(targets=[self._values_var(ast.Str(varname), ctx=ast.Store())], value=safe)
-                ]
-
+            def_name = f"qweb_t_set_{re.sub(_VARNAME_REGEX, '_', options['last_path_node'])}"
+            content = self._compile_directive_content(el, options, indent + 1) + self._flushText(options, indent + 1)
+            if content:
+                code.append(self._indent(f"def {def_name}():", indent))
+                code.extend(content)
+                expr = f"Markup(''.join({def_name}()))"
```

Comment:
does this mean the result of `t-set` is automatically considered safe? that's not right, is it?

Reply:
This is indeed the case, the value comes from the html content and is considered safe.